src/coreclr/jit/morph.cpp

```diff
@@ -9469,11 +9462,7 @@ GenTree* Compiler::fgMorphConst(GenTree* tree)
         return tree;
     }
 
-    // TODO-CQ: Do this for compCurBB->isRunRarely(). Doing that currently will
-    // guarantee slow performance for that block. Instead cache the return value
-    // of CORINFO_HELP_STRCNS and go to cache first giving reasonable perf.
-
-    if (compCurBB->bbJumpKind == BBJ_THROW)
+    if (compCurBB->isRunRarely())
```

Comment:
Can `isRunRarely` return true for methods that are rarely run based on profile data, but do not actually end up throwing exceptions?

Reply:
@jkotas As far as I understand - yes, it adds a small overhead for cases where profile is invalid (or becomes invalid over time - we don't deopt in such cases yet) - but I guess it won't be the only performance issue in such cases.

@AndyAyersMS Maybe it worth adding a flag notifying that a block was set cold by hands in code?

In theory, I can leave this condition to be `if (compCurBB->bbJumpKind == BBJ_THROW)` but unfortunately string materialization happens before we convert blocks with throw-helpers into BBJ_THROW - I can try to delay the `fgMorphConst` for GT_CNS_STR to happen after that (or even introduce a separate phase).

Reply:
@EgorBo yes, that's more or less where we are headed.

Currently we'll set zero count blocks to rare if we have profile data. The idea is to blend this with synthetic profile data which will take a more balanced view, so that thin or stale PGO profiles won't lead to overly pessimistic codegen.

With this we can control which paths are considered "truly rare" paths.

Not sure if we'll get all the above done in .NET 6. It is on the plan, but farther out.

Reply:
@AndyAyersMS what would you recommend to do now? to leave it as is (master)?

Reply:
There are other benefits to marking blocks with no return calls as BBJ_THROW earlier -- I recall wanting to do this. So perhaps leave the test as is and look into doing that?

We recognize no return callees during inlining, so perhaps as a part of `fgTailMergeThrows`?

Reply:
Will do that in a separate PR, this will address only the non-void noreturns (-5000 bytes diff)